File: product/ml/semantic/rules.py
#!/usr/bin/env python3
"""Answer-type classifier + invention_score lookup table (T042).

THIS MODULE IS THE DEFINITION. T045 ports these exact rules to Rust and
parity-tests its output against `classify_answer_type`/`invention_score`
here — do not change the rule order or the lookup table without updating
that parity test.

Deliberately stdlib-only (re, unicodedata) so it stays trivially portable
and parity-testable: no dependency on transcribe.py, numpy, pandas or the
dataset. The DIGEST.md-building code in `main()` imports those lazily,
only when actually generating the digest.

Rules operate on the caller's answer to the probe question (§ T043's
ground truth: the first caller turn after the agent's probe turn). Answer
types, checked in this priority order (first match wins):

    1. question_back      caller asks back instead of answering
    2. denial              caller states they don't have/know the thing
    3. hedge               caller is unsure / qualifies the answer
    4. assertion_numeric   answer contains >= 3 digit characters
    5. assertion_product   answer names one of the two offered products
    6. assertion_name      answer states a personal name
    7. other               none of the above (fallback)

`invention_score(answer_type)` is a fixed lookup (documented below and in
DIGEST.md): confident assertions score high (a human with no account under
discussion has nothing to confidently assert; a template-following system
tends to invent one), denial/question_back score low, hedge sits between.
"""

from __future__ import annotations

import re
import unicodedata
from typing import Optional, TypedDict

# --------------------------------------------------------------------------
# Text normalisation (kept independent from transcribe.normalize_text on
# purpose -- this module must not import transcribe.py, see module docstring)
# --------------------------------------------------------------------------


def normalize(text: str) -> str:
    """lowercase, strip accents/punctuation (keep digits), collapse whitespace."""
    text = text.lower()
    text = unicodedata.normalize("NFKD", text)
    text = "".join(c for c in text if not unicodedata.combining(c))
    text = re.sub(r"[^a-z0-9 ]+", " ", text)
    return re.sub(r"\s+", " ", text).strip()


# --------------------------------------------------------------------------
# Answer types
# --------------------------------------------------------------------------

ANSWER_TYPES = (
    "question_back",
    "denial",
    "hedge",
    "assertion_numeric",
    "assertion_product",
    "assertion_name",
    "other",
)

# Substrings checked against the *normalized* (accent/punctuation-stripped)
# text. Order within a list doesn't matter; order between categories (the
# priority list above) does.
_QUESTION_BACK_PATTERNS = (
    "cual es",
    "cuales son",
    "que es eso",
    "que es esto",
    "que es esa",
    "a que se refiere",
    "no entiendo",
    "podria repetir",
    "puede repetir",
    "como dice",
    "perdon cual",
    "disculpe cual",
)

_DENIAL_PATTERNS = (
    "no tengo",
    "no cuento con",
    "no se",
    "no lo se",
    "no existe",
    "no me aparece",
    "no manejo",
    "no cuento",
    "ninguna",
    "ningun",
    "no cual",
)

_HEDGE_PATTERNS = (
    "creo que",
    "creo qie",
    "no estoy segur",
    "no estoy muy segur",
    "me imagino",
    "mas o menos",
    "tal vez",
    "talvez",
    "quizas",
    "supongo",
    "posiblemente",
    "puede que",
)

# "nómina plus" / "crédito verde" plus the ASR spelling variants measured
# on this dataset (same variants as transcribe.PROBE_PHRASE_VARIANTS's
# "nomina" family, extended with more variants measured against caller
# answers specifically -- callers paraphrase/mishear the product names
# whisper already mangled in the agent's own probe turn).
_PRODUCT_PATTERNS = (
    "nomina plus",
    "no mina plus",
    "no mine plus",
    "no mine a plus",
    "no minea plus",
    "no me ena plus",
    "no me na plus",
    "no mi no plus",
    "nomina fluz",
    "nomina flus",
    "nominaplus",
    "credito verde",
    "credito berde",
    "creditoverde",
    "redito verde",
    "decreto verde",
    "creto verde",
)

_NAME_INTRO_RE = re.compile(
    r"\b(me llamo|mi nombre es|soy)\b\s+([A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)"
)

_DIGIT_RE = re.compile(r"\d")


def _contains_any(normalized_text: str, patterns: tuple) -> bool:
    return any(p in normalized_text for p in patterns)


def classify_answer_type(text: str) -> str:
    """One of ANSWER_TYPES for the caller's answer `text` (raw, un-normalized)."""
    if not text or not text.strip():
        return "other"

    norm = normalize(text)

    if _contains_any(norm, _QUESTION_BACK_PATTERNS):
        return "question_back"
    if _contains_any(norm, _DENIAL_PATTERNS):
        return "denial"
    if _contains_any(norm, _HEDGE_PATTERNS):
        return "hedge"
    if len(_DIGIT_RE.findall(text)) >= 3:
        return "assertion_numeric"
    if _contains_any(norm, _PRODUCT_PATTERNS):
        return "assertion_product"
    if _NAME_INTRO_RE.search(text):
        return "assertion_name"
    return "other"


def word_count(text: str) -> int:
    return len(text.split()) if text else 0


# Fixed lookup table (§ module docstring): assertion_* high (a confident
# claim about something that shouldn't exist is the invented-information
# tell Altur's brief describes), denial/question_back low (a human with
# nothing to report says so or asks what's being asked), hedge in between.
_INVENTION_SCORE = {
    "assertion_numeric": 0.90,
    "assertion_product": 0.85,
    "assertion_name": 0.80,
    "other": 0.50,
    "hedge": 0.35,
    "question_back": 0.15,
    "denial": 0.05,
}


def invention_score(answer_type: str) -> float:
    try:
        return _INVENTION_SCORE[answer_type]
    except KeyError as exc:
        raise ValueError(f"unknown answer_type: {answer_type!r}") from exc


class AnswerAnalysis(TypedDict):
    answer_type: str
    word_count: int
    response_latency: Optional[float]
    invention_score: float


def analyze_answer(text: str, response_latency: Optional[float] = None) -> AnswerAnalysis:
    """Combines classify_answer_type/word_count/invention_score into one record."""
    answer_type = classify_answer_type(text)
    return {
        "answer_type": answer_type,
        "word_count": word_count(text),
        "response_latency": response_latency,
        "invention_score": invention_score(answer_type),
    }


# --------------------------------------------------------------------------
# DIGEST.md generation (imports common/transcribe lazily -- see module docstring)
# --------------------------------------------------------------------------

RULE_TABLE_MD = """\
| answer_type | matched when (priority order, first match wins) | invention_score |
|---|---|---|
| question_back | caller asks back ("¿cuál es?", "no entiendo", "¿podría repetir?") | 0.15 |
| denial | caller states absence ("no tengo", "no cuento con", "no sé", "no existe", "no me aparece", "ninguna") | 0.05 |
| hedge | caller qualifies the answer ("creo que", "no estoy seguro", "me imagino", "más o menos", "tal vez") | 0.35 |
| assertion_numeric | answer contains >= 3 digit characters | 0.90 |
| assertion_product | answer names one of the two offered products ("nómina plus" / "crédito verde", incl. ASR spelling variants) | 0.85 |
| assertion_name | answer states a personal name ("me llamo…", "mi nombre es…", "soy…" + capitalized word) | 0.80 |
| other | none of the above | 0.50 |
"""


def _load_ground_truth_and_transcripts(semantic_dir, model: str = "base"):
    import json

    cache_dir = semantic_dir / "cache"
    gt_path = cache_dir / "probe_ground_truth.json"
    if not gt_path.is_file():
        raise FileNotFoundError(
            f"{gt_path} not found -- run transcribe.py first (it writes this file)."
        )
    with gt_path.open("r", encoding="utf-8") as f:
        gt = json.load(f)

    rows = []
    for anon_id, entry in gt["calls"].items():
        if not entry.get("probe_found") or entry.get("answer_turn_index") is None:
            continue
        transcript_path = cache_dir / "transcripts" / model / f"{anon_id}.json"
        if not transcript_path.is_file():
            continue
        with transcript_path.open("r", encoding="utf-8") as f:
            data = json.load(f)
        turns = data["turns"]
        answer_idx = entry["answer_turn_index"]
        if answer_idx >= len(turns):
            continue
        text = turns[answer_idx]["text"]
        analysis = analyze_answer(text, entry.get("response_latency"))
        analysis["is_empty"] = not text.strip()
        rows.append({"anon_id": anon_id, **analysis})
    return gt, rows


def build_digest(semantic_dir=None, model: str = "base") -> str:
    import sys
    from pathlib import Path

    if semantic_dir is None:
        semantic_dir = Path(__file__).resolve().parent
    sys.path.insert(0, str(semantic_dir.parent))
    from common.dataset import load_manifest  # noqa: E402

    gt, rows = _load_ground_truth_and_transcripts(semantic_dir, model=model)

    manifest = load_manifest()
    label_by_id = dict(zip(manifest["anon_id"], manifest["label"]))
    for row in rows:
        row["label"] = label_by_id.get(row["anon_id"])

    n_calls = gt["considered"]
    n_probe_found = gt["found"]
    n_answers = len(rows)

    counts = {t: 0 for t in ANSWER_TYPES}
    for row in rows:
        counts[row["answer_type"]] += 1

    by_label: dict = {}
    for label in ("human", "synthetic"):
        label_rows = [r for r in rows if r["label"] == label]
        n = len(label_rows)
        label_counts = {t: 0 for t in ANSWER_TYPES}
        for r in label_rows:
            label_counts[r["answer_type"]] += 1
        mean_score = (sum(r["invention_score"] for r in label_rows) / n) if n else 0.0
        by_label[label] = {"n": n, "counts": label_counts, "mean_invention_score": mean_score}

    lines = []
    lines.append("# Semantic layer digest (T042)\n")
    lines.append(
        f"Base model transcripts + probe ground truth over {n_calls} calls with a "
        f"cached base-model transcript: probe phrase found in {n_probe_found} "
        f"({n_probe_found / n_calls:.1%}) calls; a caller answer turn was "
        f"identifiable for {n_answers} ({n_answers / n_calls:.1%}) of them.\n"
    )
    lines.append("## Rule table\n")
    lines.append(RULE_TABLE_MD)

    n_empty = sum(1 for r in rows if r["is_empty"])
    lines.append(
        f"\n**Data-quality note**: {n_empty}/{n_answers} ({n_empty / n_answers:.1%}) "
        "identified answer turns transcribed as *empty text* -- \"the first caller "
        "turn after the probe\" (§ T043's fixed definition) is often a brief "
        "interjection/breath with no ASR-recognisable speech, not the caller's "
        "substantive reply. These are classified `other` (word_count=0) below, "
        "which dominates that bucket -- this is a data-quality fact about the "
        "ground truth's turn-selection rule, not a gap in the rule table itself."
    )

    lines.append("\n## answer_type distribution (all calls with an identified answer)\n")
    lines.append("| answer_type | count | share |")
    lines.append("|---|---|---|")
    for t in ANSWER_TYPES:
        share = (counts[t] / n_answers) if n_answers else 0.0
        lines.append(f"| {t} | {counts[t]} | {share:.1%} |")

    lines.append(
        "\n## answer_type distribution by manifest label -- the first honest look "
        "at whether F-23 carries signal\n"
    )
    lines.append(
        "If `invention_score` carries no signal, its mean should land close to "
        "identical for `human` and `synthetic` calls; a gap is evidence the rule "
        "table is picking up a real behavioral difference, not noise.\n"
    )
    lines.append("| label | n | " + " | ".join(ANSWER_TYPES) + " | mean invention_score |")
    lines.append("|---|---|" + "---|" * len(ANSWER_TYPES) + "---|")
    for label in ("human", "synthetic"):
        stats = by_label[label]
        n = stats["n"]
        cells = [
            f"{stats['counts'][t]} ({stats['counts'][t] / n:.0%})" if n else "0"
            for t in ANSWER_TYPES
        ]
        lines.append(
            f"| {label} | {n} | " + " | ".join(cells) + f" | {stats['mean_invention_score']:.3f} |"
        )

    human_mean = by_label["human"]["mean_invention_score"]
    synthetic_mean = by_label["synthetic"]["mean_invention_score"]
    gap = synthetic_mean - human_mean
    lines.append(
        f"\nMean invention_score gap (synthetic - human): **{gap:+.3f}** "
        f"({synthetic_mean:.3f} vs {human_mean:.3f}). "
        + (
            "Directionally consistent with the thesis (a scripted responder invents "
            "more than a human who can say \"I don't have that\") -- worth carrying "
            "into fc-2 as an optional F-23 candidate, pending a larger sample."
            if gap > 0.03
            else "No clear separation on this sample -- do not treat F-23 as a strong "
            "signal without further validation before adding it to a future contract."
        )
    )
    lines.append("")
    return "\n".join(lines)


def main() -> int:
    from pathlib import Path

    semantic_dir = Path(__file__).resolve().parent
    digest = build_digest(semantic_dir)
    out_path = semantic_dir / "DIGEST.md"
    out_path.write_text(digest, encoding="utf-8")
    print(f"wrote {out_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: product/api/src/semantic/rules.rs
//! Answer-type classifier + `invention_score` lookup table (T045), ported
//! line-by-line from `product/ml/semantic/rules.py` — THAT module is the
//! definition; this one must be parity-tested against it (see
//! `product/ml/semantic/tests/test_rules.py` and this module's own tests
//! below, mirroring the same cases). Do not change the rule priority order
//! or the lookup table here without updating `rules.py` first.
//!
//! Deliberately dependency-free (no `regex`, no `unicode-normalization`):
//! `rules.py`'s `normalize()` runs full Unicode NFKD + combining-mark
//! stripping, but the only accented characters this pipeline ever actually
//! sees are Spanish ASR output (á é í ó ú ñ ü, upper/lower) — `fold_accents`
//! below special-cases exactly that set rather than pulling in a general
//! Unicode normalization crate for one alphabet's worth of diacritics.
//!
//! Rules operate on the caller's answer to the T043 probe question (the
//! first caller turn after the detected probe turn), checked in this
//! priority order (first match wins):
//!
//! ```text
//! 1. question_back      caller asks back instead of answering
//! 2. denial              caller states they don't have/know the thing
//! 3. hedge               caller is unsure / qualifies the answer
//! 4. assertion_numeric   answer contains >= 3 digit characters
//! 5. assertion_product   answer names one of the two offered products
//! 6. assertion_name      answer states a personal name
//! 7. other               none of the above (fallback)
//! ```

/// One of the seven answer-type categories (§ module docstring priority
/// order). `Other` doubles as both "no rule matched" and "empty/whitespace
/// text", matching `rules.py::classify_answer_type`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum AnswerType {
    QuestionBack,
    Denial,
    Hedge,
    AssertionNumeric,
    AssertionProduct,
    AssertionName,
    Other,
}

impl AnswerType {
    /// `rules.py::invention_score` — fixed lookup, confident assertions
    /// score high, denial/question_back score low, hedge in between.
    pub fn invention_score(self) -> f64 {
        match self {
            AnswerType::AssertionNumeric => 0.90,
            AnswerType::AssertionProduct => 0.85,
            AnswerType::AssertionName => 0.80,
            AnswerType::Other => 0.50,
            AnswerType::Hedge => 0.35,
            AnswerType::QuestionBack => 0.15,
            AnswerType::Denial => 0.05,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            AnswerType::QuestionBack => "question_back",
            AnswerType::Denial => "denial",
            AnswerType::Hedge => "hedge",
            AnswerType::AssertionNumeric => "assertion_numeric",
            AnswerType::AssertionProduct => "assertion_product",
            AnswerType::AssertionName => "assertion_name",
            AnswerType::Other => "other",
        }
    }
}

// --------------------------------------------------------------------------
// Text normalisation (mirrors rules.py::normalize for the Spanish alphabet)
// --------------------------------------------------------------------------

fn fold_accents(c: char) -> char {
    match c {
        'á' | 'à' | 'ä' | 'â' | 'Á' | 'À' | 'Ä' | 'Â' => 'a',
        'é' | 'è' | 'ë' | 'ê' | 'É' | 'È' | 'Ë' | 'Ê' => 'e',
        'í' | 'ì' | 'ï' | 'î' | 'Í' | 'Ì' | 'Ï' | 'Î' => 'i',
        'ó' | 'ò' | 'ö' | 'ô' | 'Ó' | 'Ò' | 'Ö' | 'Ô' => 'o',
        'ú' | 'ù' | 'ü' | 'û' | 'Ú' | 'Ù' | 'Ü' | 'Û' => 'u',
        'ñ' | 'Ñ' => 'n',
        other => other,
    }
}

/// lowercase, strip accents/punctuation (keep digits), collapse whitespace.
pub fn normalize(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    let mut last_was_space = true; // suppress leading spaces
    for c in text.chars() {
        let c = fold_accents(c.to_ascii_lowercase());
        // `to_ascii_lowercase` only folds ASCII; accented uppercase is
        // handled directly by `fold_accents`'s own upper/lower arms above.
        let mapped = if c.is_ascii_lowercase() || c.is_ascii_digit() {
            c
        } else {
            ' '
        };
        if mapped == ' ' {
            if !last_was_space {
                out.push(' ');
            }
            last_was_space = true;
        } else {
            out.push(mapped);
            last_was_space = false;
        }
    }
    while out.ends_with(' ') {
        out.pop();
    }
    out
}

// --------------------------------------------------------------------------
// Pattern tables (checked as substrings of the normalized text)
// --------------------------------------------------------------------------

const QUESTION_BACK_PATTERNS: &[&str] = &[
    "cual es",
    "cuales son",
    "que es eso",
    "que es esto",
    "que es esa",
    "a que se refiere",
    "no entiendo",
    "podria repetir",
    "puede repetir",
    "como dice",
    "perdon cual",
    "disculpe cual",
];

const DENIAL_PATTERNS: &[&str] = &[
    "no tengo",
    "no cuento con",
    "no se",
    "no lo se",
    "no existe",
    "no me aparece",
    "no manejo",
    "no cuento",
    "ninguna",
    "ningun",
    "no cual",
];

const HEDGE_PATTERNS: &[&str] = &[
    "creo que",
    "creo qie",
    "no estoy segur",
    "no estoy muy segur",
    "me imagino",
    "mas o menos",
    "tal vez",
    "talvez",
    "quizas",
    "supongo",
    "posiblemente",
    "puede que",
];

const PRODUCT_PATTERNS: &[&str] = &[
    "nomina plus",
    "no mina plus",
    "no mine plus",
    "no mine a plus",
    "no minea plus",
    "no me ena plus",
    "no me na plus",
    "no mi no plus",
    "nomina fluz",
    "nomina flus",
    "nominaplus",
    "credito verde",
    "credito berde",
    "creditoverde",
    "redito verde",
    "decreto verde",
    "creto verde",
];

const NAME_INTROS: &[&str] = &["me llamo", "mi nombre es", "soy"];

fn contains_any(normalized: &str, patterns: &[&str]) -> bool {
    patterns.iter().any(|p| normalized.contains(p))
}

/// `rules.py::_NAME_INTRO_RE`: `\b(me llamo|mi nombre es|soy)\b\s+([A-Z...][a-z...]+)`
/// against the RAW (un-normalized) text — an intro phrase (case-insensitive,
/// whole-word) immediately followed by whitespace and a capitalized word.
fn contains_name_intro(text: &str) -> bool {
    let chars: Vec<char> = text.chars().collect();
    let lower: String = chars.iter().map(|c| c.to_ascii_lowercase()).collect();
    let lower_chars: Vec<char> = lower.chars().collect();

    let is_word_char = |c: char| c.is_alphanumeric() || c == '_';

    for intro in NAME_INTROS {
        let intro_chars: Vec<char> = intro.chars().collect();
        let n = intro_chars.len();
        if n > lower_chars.len() {
            continue;
        }
        for start in 0..=(lower_chars.len() - n) {
            if lower_chars[start..start + n] != intro_chars[..] {
                continue;
            }
            // whole-word: char before start (if any) and char after the
            // match (if any, before whitespace) must not be word chars.
            let before_ok = start == 0 || !is_word_char(lower_chars[start - 1]);
            let end = start + n;
            let after_ok = end == lower_chars.len() || !is_word_char(lower_chars[end]);
            if !before_ok || !after_ok {
                continue;
            }
            // skip one-or-more whitespace
            let mut i = end;
            let mut saw_space = false;
            while i < chars.len() && chars[i].is_whitespace() {
                i += 1;
                saw_space = true;
            }
            if !saw_space || i >= chars.len() {
                continue;
            }
            // first char of the following word must be uppercase, and the
            // word must contain at least one lowercase letter after it
            // (mirrors `[A-Z...][a-z...]+`: one or more lowercase letters).
            let first = chars[i];
            if !first.is_uppercase() {
                continue;
            }
            if i + 1 < chars.len() && chars[i + 1].is_lowercase() {
                return true;
            }
        }
    }
    false
}

fn count_digits(text: &str) -> usize {
    text.chars().filter(|c| c.is_ascii_digit()).count()
}

/// `rules.py::classify_answer_type` — one of [`AnswerType`] for the
/// caller's raw (un-normalized) answer `text`.
pub fn classify_answer_type(text: &str) -> AnswerType {
    if text.trim().is_empty() {
        return AnswerType::Other;
    }

    let norm = normalize(text);

    if contains_any(&norm, QUESTION_BACK_PATTERNS) {
        return AnswerType::QuestionBack;
    }
    if contains_any(&norm, DENIAL_PATTERNS) {
        return AnswerType::Denial;
    }
    if contains_any(&norm, HEDGE_PATTERNS) {
        return AnswerType::Hedge;
    }
    if count_digits(text) >= 3 {
        return AnswerType::AssertionNumeric;
    }
    if contains_any(&norm, PRODUCT_PATTERNS) {
        return AnswerType::AssertionProduct;
    }
    if contains_name_intro(text) {
        return AnswerType::AssertionName;
    }
    AnswerType::Other
}

/// `rules.py::analyze_answer`'s core (`word_count`/`response_latency` are
/// caller concerns in this port, not part of the parity-tested surface).
pub fn analyze_answer(text: &str) -> (AnswerType, f64) {
    let answer_type = classify_answer_type(text);
    (answer_type, answer_type.invention_score())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_or_whitespace_is_other() {
        assert_eq!(classify_answer_type(""), AnswerType::Other);
        assert_eq!(classify_answer_type("   "), AnswerType::Other);
    }

    #[test]
    fn question_back_matches() {
        assert_eq!(classify_answer_type("¿Cuál es?"), AnswerType::QuestionBack);
        assert_eq!(classify_answer_type("No entiendo la pregunta"), AnswerType::QuestionBack);
        assert_eq!(classify_answer_type("¿Podría repetir por favor?"), AnswerType::QuestionBack);
    }

    #[test]
    fn denial_matches() {
        assert_eq!(classify_answer_type("No tengo esa cuenta"), AnswerType::Denial);
        assert_eq!(classify_answer_type("No sé de qué me habla"), AnswerType::Denial);
        assert_eq!(classify_answer_type("Ninguna de las dos"), AnswerType::Denial);
    }

    #[test]
    fn hedge_matches() {
        assert_eq!(classify_answer_type("Creo que es la de nómina"), AnswerType::Hedge);
        assert_eq!(classify_answer_type("No estoy segura, tal vez"), AnswerType::Hedge);
    }

    #[test]
    fn assertion_numeric_needs_at_least_three_digits() {
        assert_eq!(classify_answer_type("Es la cuenta 12"), AnswerType::Other);
        assert_eq!(classify_answer_type("Es la cuenta 1234"), AnswerType::AssertionNumeric);
    }

    #[test]
    fn assertion_product_matches_spelling_variants() {
        assert_eq!(classify_answer_type("Es la de crédito verde"), AnswerType::AssertionProduct);
        assert_eq!(classify_answer_type("Sí, la no mina plus"), AnswerType::AssertionProduct);
    }

    #[test]
    fn assertion_name_matches_intro_plus_capitalized_word() {
        assert_eq!(classify_answer_type("Me llamo Roberto"), AnswerType::AssertionName);
        assert_eq!(classify_answer_type("Mi nombre es Ana"), AnswerType::AssertionName);
        assert_eq!(classify_answer_type("Soy Carlos"), AnswerType::AssertionName);
    }

    #[test]
    fn assertion_name_requires_capitalized_word_immediately_after() {
        // lowercase word after the intro should NOT count as a name.
        assert_eq!(classify_answer_type("soy yo"), AnswerType::Other);
    }

    #[test]
    fn other_is_the_fallback() {
        assert_eq!(classify_answer_type("Buenos días, gracias"), AnswerType::Other);
    }

    #[test]
    fn priority_order_question_back_beats_denial() {
        // Contains both a denial-ish and question-back-ish substring;
        // question_back must win since it's checked first.
        assert_eq!(classify_answer_type("No entiendo, no tengo idea"), AnswerType::QuestionBack);
    }

    #[test]
    fn invention_score_lookup_matches_the_frozen_table() {
        assert_eq!(AnswerType::AssertionNumeric.invention_score(), 0.90);
        assert_eq!(AnswerType::AssertionProduct.invention_score(), 0.85);
        assert_eq!(AnswerType::AssertionName.invention_score(), 0.80);
        assert_eq!(AnswerType::Other.invention_score(), 0.50);
        assert_eq!(AnswerType::Hedge.invention_score(), 0.35);
        assert_eq!(AnswerType::QuestionBack.invention_score(), 0.15);
        assert_eq!(AnswerType::Denial.invention_score(), 0.05);
    }

    #[test]
    fn normalize_strips_accents_punctuation_and_collapses_whitespace() {
        assert_eq!(normalize("¡Crédito   Verde!"), "credito verde");
        assert_eq!(normalize("Nómina  Plus?"), "nomina plus");
    }

    #[test]
    fn as_str_round_trips_all_variants() {
        for (t, s) in [
            (AnswerType::QuestionBack, "question_back"),
            (AnswerType::Denial, "denial"),
            (AnswerType::Hedge, "hedge"),
            (AnswerType::AssertionNumeric, "assertion_numeric"),
            (AnswerType::AssertionProduct, "assertion_product"),
            (AnswerType::AssertionName, "assertion_name"),
            (AnswerType::Other, "other"),
        ] {
            assert_eq!(t.as_str(), s);
        }
    }
}
